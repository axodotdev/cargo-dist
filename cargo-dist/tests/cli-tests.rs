use std::{
    path::PathBuf,
    process::{Command, Output, Stdio},
};

#[cfg(unix)]
use std::{fs::File, os::unix::fs::PermissionsExt};

static BIN: &str = env!("CARGO_BIN_EXE_dist");
const ENV_RUIN_ME: &str = "RUIN_MY_COMPUTER_WITH_INSTALLERS";

#[allow(dead_code)]
mod gallery;
use axoasset::LocalAsset;
use axoprocess::Cmd;
use axoupdater::{test::helpers::RuntestArgs, AxoUpdater, ReleaseSourceType};
use camino::Utf8PathBuf;
use gallery::*;

fn format_outputs(output: &Output) -> String {
    let stdout = std::str::from_utf8(&output.stdout).unwrap();
    let stderr = std::str::from_utf8(&output.stderr).unwrap();
    format!("stdout:\n{stdout}\nstderr:\n{stderr}")
}

#[test]
fn test_version() {
    let output = Command::new(BIN)
        .arg("dist")
        .arg("-V")
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .output()
        .unwrap();

    let stdout = String::from_utf8(output.stdout).unwrap();
    let stderr = String::from_utf8(output.stderr).unwrap();

    assert!(output.status.success(), "{}", stderr);
    assert_eq!(stderr, "");

    let (name, ver) = stdout.split_once(' ').unwrap();
    assert_eq!(name, "cargo-dist");
    let mut ver_parts = ver.trim().split('.');
    ver_parts.next().unwrap().parse::<u8>().unwrap();
    ver_parts.next().unwrap().parse::<u8>().unwrap();
    let last = ver_parts.next().unwrap();
    if let Some((last, _prerelease)) = last.split_once('-') {
        last.parse::<u8>().unwrap();
        if let Some(build) = ver_parts.next() {
            build.parse::<u8>().unwrap();
        }
    } else {
        last.parse::<u8>().unwrap();
    }
    assert!(ver_parts.next().is_none());
}

#[test]
fn test_long_help() {
    let output = Command::new(BIN)
        .arg("dist")
        .arg("--help")
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .output()
        .unwrap();

    snapshot_settings().bind(|| {
        insta::assert_snapshot!("long-help", format_outputs(&output));
    });
    assert!(output.status.success(), "{}", output.status);
}

#[test]
fn test_short_help() {
    let output = Command::new(BIN)
        .arg("dist")
        .arg("-h")
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .output()
        .unwrap();

    snapshot_settings().bind(|| {
        insta::assert_snapshot!("short-help", format_outputs(&output));
    });
    assert!(output.status.success(), "{}", output.status);
}

#[test]
fn test_manifest() {
    let output = Command::new(BIN)
        .arg("dist")
        .arg("manifest")
        .arg("--artifacts=lies")
        .arg("--no-local-paths")
        .arg("--allow-dirty")
        .arg("--output-format=json")
        .arg("--verbose=error")
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .output()
        .unwrap();

    // We don't want this to churn every time we do a version bump
    snapshot_settings_with_dist_manifest_filter().bind(|| {
        insta::assert_snapshot!(format_outputs(&output));
    });

    assert!(output.status.success(), "{}", output.status);
}

#[test]
fn test_lib_manifest() {
    let version = std::env!("CARGO_PKG_VERSION");
    let output = Command::new(BIN)
        .arg("dist")
        .arg("manifest")
        .arg("--artifacts=all")
        .arg("--no-local-paths")
        .arg("--allow-dirty")
        .arg("--output-format=json")
        .arg("--verbose=error")
        .arg("--tag")
        .arg(format!("cargo-dist-schema-v{}", version))
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .output()
        .unwrap();

    // We don't want this to churn every time we do a version bump
    snapshot_settings_with_dist_manifest_filter().bind(|| {
        insta::assert_snapshot!(format_outputs(&output));
    });

    assert!(output.status.success(), "{}", output.status);
}

#[test]
fn test_lib_manifest_slash() {
    let version = std::env!("CARGO_PKG_VERSION");
    let output = Command::new(BIN)
        .arg("dist")
        .arg("manifest")
        .arg("--artifacts=all")
        .arg("--no-local-paths")
        .arg("--allow-dirty")
        .arg("--output-format=json")
        .arg("--verbose=error")
        .arg("--tag")
        .arg(format!("cargo-dist-schema/v{}", version))
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .output()
        .unwrap();

    // We don't want this to churn every time we do a version bump
    snapshot_settings_with_dist_manifest_filter().bind(|| {
        insta::assert_snapshot!(format_outputs(&output));
    });

    assert!(output.status.success(), "{}", output.status);
}

#[test]
fn test_error_manifest() {
    let output = Command::new(BIN)
        .arg("dist")
        .arg("manifest")
        .arg("--artifacts=all")
        .arg("--no-local-paths")
        .arg("--allow-dirty")
        .arg("--output-format=json")
        .arg("--verbose=error")
        .arg("--tag=v0.0.0")
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .output()
        .unwrap();

    // We don't want this to churn every time we do a version bump
    snapshot_settings_with_dist_manifest_filter().bind(|| {
        insta::assert_snapshot!(format_outputs(&output));
    });

    assert!(!output.status.success(), "{}", output.status);
}

#[test]
fn test_markdown_help() {
    let output = Command::new(BIN)
        .arg("dist")
        .arg("help-markdown")
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .output()
        .unwrap();

    snapshot_settings().bind(|| {
        insta::assert_snapshot!("markdown-help", format_outputs(&output));
    });
    assert!(output.status.success(), "{}", output.status);
}

fn generate_installer(version: &axotag::Version, release_type: ReleaseSourceType) -> Utf8PathBuf {
    let tools = Tools::default();

    // On Windows, we need a debug build of cargo-dist present
    // in order to be able to build the global artifacts. This is
    // due to the extra-artifacts, which we don't yet have a CLI
    // option to turn off during `cargo dist build`.
    Cmd::new("cargo", "run debug build")
        .arg("build")
        .status()
        .unwrap();

    // First update the cargo-dist version so we can safely `cargo dist build`
    tools
        .cargo_dist
        .output_checked(|cmd| cmd.arg("dist").arg("init").arg("--yes"))
        .unwrap();

    // Now we run cargo-dist to generate an installer for the current version
    tools
        .cargo_dist
        .output_checked(|cmd| cmd.arg("dist").arg("build").arg("--artifacts=global"))
        .unwrap();

    let ext = if cfg!(windows) { ".ps1" } else { ".sh" };
    let root = env!("CARGO_MANIFEST_DIR");
    let installer_path = Utf8PathBuf::from(root)
        .parent()
        .unwrap()
        .join("target")
        .join("distrib")
        .join(format!("cargo-dist-installer{ext}"));
    let installer_string = std::fs::read_to_string(&installer_path).unwrap();

    let installer_url = match release_type {
        ReleaseSourceType::Axo => {
            format!("https://axodotdev.artifacts.axodotdev.host/cargo-dist/v{version}",)
        }
        ReleaseSourceType::GitHub => {
            format!("https://github.com/axodotdev/cargo-dist/releases/download/v{version}",)
        }
    };

    let installer_string = installer_string
        .replace(env!("CARGO_PKG_VERSION"), &version.to_string())
        .replace(
            "https://fake.axo.dev/faker/cargo-dist/fake-id-do-not-upload",
            &installer_url,
        );

    #[cfg(unix)]
    {
        let installer_file = File::create(&installer_path).unwrap();
        let mut perms = installer_file.metadata().unwrap().permissions();
        perms.set_mode(0o744);
        installer_file.set_permissions(perms).unwrap();
    }

    LocalAsset::write_new(&installer_string, &installer_path).unwrap();

    installer_path
}

#[test]
fn test_self_update() {
    // Only do this if RUIN_MY_COMPUTER_WITH_INSTALLERS is set
    if std::env::var(ENV_RUIN_ME)
        .map(|s| s == "selfupdate" || s == "all")
        .unwrap_or(false)
    {
        std::env::remove_var("XDG_CONFIG_HOME");

        let mut args = RuntestArgs {
            app_name: "cargo-dist".to_owned(),
            package: "cargo-dist".to_owned(),
            owner: "axodotdev".to_owned(),
            bin: PathBuf::from(BIN),
            binaries: vec!["dist".to_owned()],
            args: vec![
                "dist".to_owned(),
                "selfupdate".to_owned(),
                // init includes interactive components, so we
                // can't safely run it within a noninteractive test
                "--skip-init".to_owned(),
            ],
            release_type: ReleaseSourceType::GitHub,
        };

        // First run with GitHub
        let installed_bin = axoupdater::test::helpers::perform_runtest(&args);
        assert!(installed_bin.exists());
        let status = Command::new(&installed_bin)
            .arg("--version")
            .status()
            .expect("binary didn't exist or --version returned nonzero");
        assert!(status.success());

        // Remove the installed binary before running the next test
        std::fs::remove_file(installed_bin).unwrap();

        // Next two runtests: like the above, but we produce
        // new installers so that we test against the latest installer
        // code in this PR instead of the installers that were generated
        // for the releases we're fetching.
        let tokio = tokio::runtime::Builder::new_current_thread()
            .worker_threads(1)
            .max_blocking_threads(128)
            .enable_all()
            .build()
            .expect("Initializing tokio runtime failed");

        let mut updater = AxoUpdater::new_for("cargo-dist");
        updater.set_release_source(axoupdater::ReleaseSource {
            release_type: ReleaseSourceType::GitHub,
            owner: "axodotdev".to_owned(),
            name: "cargo-dist".to_owned(),
            app_name: "cargo-dist".to_owned(),
        });
        // This is the new version that we'll create alternate installers for.
        let new_version = tokio
            .block_on(updater.query_new_version())
            .unwrap()
            .unwrap();

        let installer_path = generate_installer(new_version, ReleaseSourceType::GitHub);

        // OK now, finally, we have an installer at `installer_path`
        // with URLs pointing at the exact version we want to test.
        // Point cargo-dist at it.
        std::env::set_var("CARGO_DIST_USE_INSTALLER_AT_PATH", installer_path);

        args.release_type = ReleaseSourceType::GitHub;
        let installed_bin = axoupdater::test::helpers::perform_runtest(&args);
        assert!(installed_bin.exists());
        let status = Command::new(&installed_bin)
            .arg("--version")
            .status()
            .expect("binary didn't exist or --version returned nonzero");
        assert!(status.success());
    }
}
