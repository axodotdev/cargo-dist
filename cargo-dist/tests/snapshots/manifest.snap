---
source: cargo-dist/tests/cli-tests.rs
expression: format_outputs(&output)
---
stdout:
{
  "dist_version": "1.0.0-FAKEVERSION",
  "announcement_tag": "CENSORED",
  "announcement_tag_is_implicit": true,
  "announcement_is_prerelease": "CENSORED"
  "announcement_title": "CENSORED"
  "announcement_changelog": "CENSORED"
  "announcement_github_body": "CENSORED"
  "releases": [
    {
      "app_name": "cargo-dist",
      "app_version": "1.0.0-FAKEVERSION",
      "env": {
        "install_dir_env_var": "CARGO_DIST_INSTALL_DIR",
        "unmanaged_dir_env_var": "CARGO_DIST_UNMANAGED_INSTALL",
        "disable_update_env_var": "CARGO_DIST_DISABLE_UPDATE",
        "no_modify_path_env_var": "CARGO_DIST_NO_MODIFY_PATH",
        "print_quiet_env_var": "CARGO_DIST_PRINT_QUIET",
        "print_verbose_env_var": "CARGO_DIST_PRINT_VERBOSE",
        "download_url_env_var": "CARGO_DIST_DOWNLOAD_URL",
        "github_base_url_env_var": "CARGO_DIST_INSTALLER_GITHUB_BASE_URL",
        "ghe_base_url_env_var": "CARGO_DIST_INSTALLER_GHE_BASE_URL",
        "github_token_env_var": "CARGO_DIST_GITHUB_TOKEN"
      },
      "display_name": "cargo-dist",
      "display": true,
      "artifacts": [
        "source.tar.gz",
        "source.tar.gz.sha256",
        "dist-manifest-schema.json",
        "cargo-dist-installer.sh",
        "cargo-dist-installer.ps1",
        "cargo-dist.rb",
        "cargo-dist-npm-package.tar.gz",
        "sha256.sum",
        "cargo-dist-aarch64-apple-darwin.tar.xz",
        "cargo-dist-aarch64-apple-darwin.tar.xz.sha256",
        "cargo-dist-aarch64-unknown-linux-gnu.tar.xz",
        "cargo-dist-aarch64-unknown-linux-gnu.tar.xz.sha256",
        "cargo-dist-aarch64-unknown-linux-musl.tar.xz",
        "cargo-dist-aarch64-unknown-linux-musl.tar.xz.sha256",
        "cargo-dist-x86_64-apple-darwin.tar.xz",
        "cargo-dist-x86_64-apple-darwin.tar.xz.sha256",
        "cargo-dist-x86_64-pc-windows-msvc.zip",
        "cargo-dist-x86_64-pc-windows-msvc.zip.sha256",
        "cargo-dist-x86_64-unknown-linux-gnu.tar.xz",
        "cargo-dist-x86_64-unknown-linux-gnu.tar.xz.sha256",
        "cargo-dist-x86_64-unknown-linux-musl.tar.xz",
        "cargo-dist-x86_64-unknown-linux-musl.tar.xz.sha256"
      ],
      "hosting": {
        "github": {
          "artifact_base_url": "https://github.com",
          "artifact_download_path": "/axodotdev/cargo-dist/releases/download/v1.0.0-FAKEVERSION",
          "owner": "axodotdev",
          "repo": "cargo-dist"
        }
      }
    }
  ],
  "artifacts": {
    "cargo-dist-aarch64-apple-darwin.tar.xz": {
      "name": "cargo-dist-aarch64-apple-darwin.tar.xz",
      "kind": "executable-zip",
      "target_triples": [
        "aarch64-apple-darwin"
      ],
      "assets": [
        {
          "name": "CHANGELOG.md",
          "path": "CHANGELOG.md",
          "kind": "changelog"
        },
        {
          "name": "LICENSE-APACHE",
          "path": "LICENSE-APACHE",
          "kind": "license"
        },
        {
          "name": "LICENSE-MIT",
          "path": "LICENSE-MIT",
          "kind": "license"
        },
        {
          "name": "README.md",
          "path": "README.md",
          "kind": "readme"
        },
        {
          "id": "cargo-dist-aarch64-apple-darwin-exe-dist",
          "name": "dist",
          "path": "dist",
          "kind": "executable"
        }
      ],
      "checksum": "cargo-dist-aarch64-apple-darwin.tar.xz.sha256"
    },
    "cargo-dist-aarch64-apple-darwin.tar.xz.sha256": {
      "name": "cargo-dist-aarch64-apple-darwin.tar.xz.sha256",
      "kind": "checksum",
      "target_triples": [
        "aarch64-apple-darwin"
      ]
    },
    "cargo-dist-aarch64-unknown-linux-gnu.tar.xz": {
      "name": "cargo-dist-aarch64-unknown-linux-gnu.tar.xz",
      "kind": "executable-zip",
      "target_triples": [
        "aarch64-unknown-linux-gnu"
      ],
      "assets": [
        {
          "name": "CHANGELOG.md",
          "path": "CHANGELOG.md",
          "kind": "changelog"
        },
        {
          "name": "LICENSE-APACHE",
          "path": "LICENSE-APACHE",
          "kind": "license"
        },
        {
          "name": "LICENSE-MIT",
          "path": "LICENSE-MIT",
          "kind": "license"
        },
        {
          "name": "README.md",
          "path": "README.md",
          "kind": "readme"
        },
        {
          "id": "cargo-dist-aarch64-unknown-linux-gnu-exe-dist",
          "name": "dist",
          "path": "dist",
          "kind": "executable"
        }
      ],
      "checksum": "cargo-dist-aarch64-unknown-linux-gnu.tar.xz.sha256"
    },
    "cargo-dist-aarch64-unknown-linux-gnu.tar.xz.sha256": {
      "name": "cargo-dist-aarch64-unknown-linux-gnu.tar.xz.sha256",
      "kind": "checksum",
      "target_triples": [
        "aarch64-unknown-linux-gnu"
      ]
    },
    "cargo-dist-aarch64-unknown-linux-musl.tar.xz": {
      "name": "cargo-dist-aarch64-unknown-linux-musl.tar.xz",
      "kind": "executable-zip",
      "target_triples": [
        "aarch64-unknown-linux-musl"
      ],
      "assets": [
        {
          "name": "CHANGELOG.md",
          "path": "CHANGELOG.md",
          "kind": "changelog"
        },
        {
          "name": "LICENSE-APACHE",
          "path": "LICENSE-APACHE",
          "kind": "license"
        },
        {
          "name": "LICENSE-MIT",
          "path": "LICENSE-MIT",
          "kind": "license"
        },
        {
          "name": "README.md",
          "path": "README.md",
          "kind": "readme"
        },
        {
          "id": "cargo-dist-aarch64-unknown-linux-musl-exe-dist",
          "name": "dist",
          "path": "dist",
          "kind": "executable"
        }
      ],
      "checksum": "cargo-dist-aarch64-unknown-linux-musl.tar.xz.sha256"
    },
    "cargo-dist-aarch64-unknown-linux-musl.tar.xz.sha256": {
      "name": "cargo-dist-aarch64-unknown-linux-musl.tar.xz.sha256",
      "kind": "checksum",
      "target_triples": [
        "aarch64-unknown-linux-musl"
      ]
    },
    "cargo-dist-installer.ps1": {
      "name": "cargo-dist-installer.ps1",
      "kind": "installer",
      "target_triples": [
        "aarch64-pc-windows-msvc",
        "x86_64-pc-windows-gnu",
        "x86_64-pc-windows-msvc"
      ],
      "install_hint": "powershell -ExecutionPolicy Bypass -c \"irm https://github.com/axodotdev/cargo-dist/releases/download/v1.0.0-FAKEVERSION/cargo-dist-installer.ps1 | iex\"",
      "description": "Install prebuilt binaries via powershell script"
    },
    "cargo-dist-installer.sh": {
      "name": "cargo-dist-installer.sh",
      "kind": "installer",
      "target_triples": [
        "aarch64-apple-darwin",
        "aarch64-unknown-linux-gnu",
        "aarch64-unknown-linux-musl-dynamic",
        "aarch64-unknown-linux-musl-static",
        "x86_64-apple-darwin",
        "x86_64-pc-windows-gnu",
        "x86_64-unknown-linux-gnu",
        "x86_64-unknown-linux-musl-dynamic",
        "x86_64-unknown-linux-musl-static"
      ],
      "install_hint": "curl --proto '=https' --tlsv1.2 -LsSf https://github.com/axodotdev/cargo-dist/releases/download/v1.0.0-FAKEVERSION/cargo-dist-installer.sh | sh",
      "description": "Install prebuilt binaries via shell script"
    },
    "cargo-dist-npm-package.tar.gz": {
      "name": "cargo-dist-npm-package.tar.gz",
      "kind": "installer",
      "target_triples": [
        "aarch64-apple-darwin",
        "aarch64-pc-windows-msvc",
        "aarch64-unknown-linux-gnu",
        "aarch64-unknown-linux-musl-dynamic",
        "aarch64-unknown-linux-musl-static",
        "x86_64-apple-darwin",
        "x86_64-pc-windows-gnu",
        "x86_64-pc-windows-msvc",
        "x86_64-unknown-linux-gnu",
        "x86_64-unknown-linux-musl-dynamic",
        "x86_64-unknown-linux-musl-static"
      ],
      "assets": [
        {
          "name": ".gitignore",
          "path": ".gitignore",
          "kind": "unknown"
        },
        {
          "name": "CHANGELOG.md",
          "path": "CHANGELOG.md",
          "kind": "changelog"
        },
        {
          "name": "LICENSE-APACHE",
          "path": "LICENSE-APACHE",
          "kind": "license"
        },
        {
          "name": "LICENSE-MIT",
          "path": "LICENSE-MIT",
          "kind": "license"
        },
        {
          "name": "README.md",
          "path": "README.md",
          "kind": "readme"
        },
        {
          "name": "binary-install.js",
          "path": "binary-install.js",
          "kind": "unknown"
        },
        {
          "name": "binary.js",
          "path": "binary.js",
          "kind": "unknown"
        },
        {
          "name": "install.js",
          "path": "install.js",
          "kind": "unknown"
        },
        {
          "name": "npm-shrinkwrap.json",
          "path": "npm-shrinkwrap.json",
          "kind": "unknown"
        },
        {
          "name": "package.json",
          "path": "package.json",
          "kind": "unknown"
        },
        {
          "name": "run.js",
          "path": "run.js",
          "kind": "unknown"
        }
      ],
      "install_hint": "npm install @axodotdev/dist@1.0.0-FAKEVERSION",
      "description": "Install prebuilt binaries into your npm project"
    },
    "cargo-dist-x86_64-apple-darwin.tar.xz": {
      "name": "cargo-dist-x86_64-apple-darwin.tar.xz",
      "kind": "executable-zip",
      "target_triples": [
        "x86_64-apple-darwin"
      ],
      "assets": [
        {
          "name": "CHANGELOG.md",
          "path": "CHANGELOG.md",
          "kind": "changelog"
        },
        {
          "name": "LICENSE-APACHE",
          "path": "LICENSE-APACHE",
          "kind": "license"
        },
        {
          "name": "LICENSE-MIT",
          "path": "LICENSE-MIT",
          "kind": "license"
        },
        {
          "name": "README.md",
          "path": "README.md",
          "kind": "readme"
        },
        {
          "id": "cargo-dist-x86_64-apple-darwin-exe-dist",
          "name": "dist",
          "path": "dist",
          "kind": "executable"
        }
      ],
      "checksum": "cargo-dist-x86_64-apple-darwin.tar.xz.sha256"
    },
    "cargo-dist-x86_64-apple-darwin.tar.xz.sha256": {
      "name": "cargo-dist-x86_64-apple-darwin.tar.xz.sha256",
      "kind": "checksum",
      "target_triples": [
        "x86_64-apple-darwin"
      ]
    },
    "cargo-dist-x86_64-pc-windows-msvc.zip": {
      "name": "cargo-dist-x86_64-pc-windows-msvc.zip",
      "kind": "executable-zip",
      "target_triples": [
        "x86_64-pc-windows-msvc"
      ],
      "assets": [
        {
          "name": "CHANGELOG.md",
          "path": "CHANGELOG.md",
          "kind": "changelog"
        },
        {
          "name": "LICENSE-APACHE",
          "path": "LICENSE-APACHE",
          "kind": "license"
        },
        {
          "name": "LICENSE-MIT",
          "path": "LICENSE-MIT",
          "kind": "license"
        },
        {
          "name": "README.md",
          "path": "README.md",
          "kind": "readme"
        },
        {
          "id": "cargo-dist-x86_64-pc-windows-msvc-exe-dist",
          "name": "dist",
          "path": "dist.exe",
          "kind": "executable"
        }
      ],
      "checksum": "cargo-dist-x86_64-pc-windows-msvc.zip.sha256"
    },
    "cargo-dist-x86_64-pc-windows-msvc.zip.sha256": {
      "name": "cargo-dist-x86_64-pc-windows-msvc.zip.sha256",
      "kind": "checksum",
      "target_triples": [
        "x86_64-pc-windows-msvc"
      ]
    },
    "cargo-dist-x86_64-unknown-linux-gnu.tar.xz": {
      "name": "cargo-dist-x86_64-unknown-linux-gnu.tar.xz",
      "kind": "executable-zip",
      "target_triples": [
        "x86_64-unknown-linux-gnu"
      ],
      "assets": [
        {
          "name": "CHANGELOG.md",
          "path": "CHANGELOG.md",
          "kind": "changelog"
        },
        {
          "name": "LICENSE-APACHE",
          "path": "LICENSE-APACHE",
          "kind": "license"
        },
        {
          "name": "LICENSE-MIT",
          "path": "LICENSE-MIT",
          "kind": "license"
        },
        {
          "name": "README.md",
          "path": "README.md",
          "kind": "readme"
        },
        {
          "id": "cargo-dist-x86_64-unknown-linux-gnu-exe-dist",
          "name": "dist",
          "path": "dist",
          "kind": "executable"
        }
      ],
      "checksum": "cargo-dist-x86_64-unknown-linux-gnu.tar.xz.sha256"
    },
    "cargo-dist-x86_64-unknown-linux-gnu.tar.xz.sha256": {
      "name": "cargo-dist-x86_64-unknown-linux-gnu.tar.xz.sha256",
      "kind": "checksum",
      "target_triples": [
        "x86_64-unknown-linux-gnu"
      ]
    },
    "cargo-dist-x86_64-unknown-linux-musl.tar.xz": {
      "name": "cargo-dist-x86_64-unknown-linux-musl.tar.xz",
      "kind": "executable-zip",
      "target_triples": [
        "x86_64-unknown-linux-musl"
      ],
      "assets": [
        {
          "name": "CHANGELOG.md",
          "path": "CHANGELOG.md",
          "kind": "changelog"
        },
        {
          "name": "LICENSE-APACHE",
          "path": "LICENSE-APACHE",
          "kind": "license"
        },
        {
          "name": "LICENSE-MIT",
          "path": "LICENSE-MIT",
          "kind": "license"
        },
        {
          "name": "README.md",
          "path": "README.md",
          "kind": "readme"
        },
        {
          "id": "cargo-dist-x86_64-unknown-linux-musl-exe-dist",
          "name": "dist",
          "path": "dist",
          "kind": "executable"
        }
      ],
      "checksum": "cargo-dist-x86_64-unknown-linux-musl.tar.xz.sha256"
    },
    "cargo-dist-x86_64-unknown-linux-musl.tar.xz.sha256": {
      "name": "cargo-dist-x86_64-unknown-linux-musl.tar.xz.sha256",
      "kind": "checksum",
      "target_triples": [
        "x86_64-unknown-linux-musl"
      ]
    },
    "cargo-dist.rb": {
      "name": "cargo-dist.rb",
      "kind": "installer",
      "target_triples": [
        "aarch64-apple-darwin",
        "aarch64-unknown-linux-gnu",
        "aarch64-unknown-linux-musl-dynamic",
        "aarch64-unknown-linux-musl-static",
        "x86_64-apple-darwin",
        "x86_64-pc-windows-gnu",
        "x86_64-unknown-linux-gnu",
        "x86_64-unknown-linux-musl-dynamic",
        "x86_64-unknown-linux-musl-static"
      ],
      "install_hint": "brew install axodotdev/tap/cargo-dist",
      "description": "Install prebuilt binaries via Homebrew"
    },
    "dist-manifest-schema.json": {
      "name": "dist-manifest-schema.json",
      "kind": "extra-artifact"
    },
    "sha256.sum": {
      "name": "sha256.sum",
      "kind": "unified-checksum"
    },
    "source.tar.gz": {
      "name": "source.tar.gz",
      "kind": "source-tarball",
      "checksum": "source.tar.gz.sha256"
    },
    "source.tar.gz.sha256": {
      "name": "source.tar.gz.sha256",
      "kind": "checksum"
    }
  },
  "systems": {
    "plan:lies:": {
      "id": "plan:lies:",
      "cargo_version_line": "CENSORED"
      "build_environment": "indeterminate"
    }
  },
  "publish_prereleases": false,
  "force_latest": false,
  "ci": {
    "github": {
      "artifacts_matrix": {
        "include": [
          {
            "runner": "macos-14",
            "host": "aarch64-apple-darwin",
            "install_dist": {
              "shell": "sh",
              "run": "curl --proto '=https' --tlsv1.2 -LsSf https://github.com/axodotdev/cargo-dist/releases/download/v1.0.0-FAKEVERSION/cargo-dist-installer.sh | sh"
            },
            "dist_args": "--artifacts=local --target=aarch64-apple-darwin",
            "targets": [
              "aarch64-apple-darwin"
            ],
            "cache_provider": "github"
          },
          {
            "runner": "ubuntu-22.04",
            "host": "x86_64-unknown-linux-gnu",
            "container": {
              "image": "quay.io/pypa/manylinux_2_28_x86_64",
              "host": "x86_64-unknown-linux-musl",
              "package_manager": null
            },
            "install_dist": {
              "shell": "sh",
              "run": "curl --proto '=https' --tlsv1.2 -LsSf https://github.com/axodotdev/cargo-dist/releases/download/v1.0.0-FAKEVERSION/cargo-dist-installer.sh | sh"
            },
            "dist_args": "--artifacts=local --target=aarch64-unknown-linux-gnu",
            "targets": [
              "aarch64-unknown-linux-gnu"
            ],
            "packages_install": "if ! command -v cargo-zigbuild > /dev/null 2>&1; then\n  if ! command -v pip3 > /dev/null 2>&1; then\n    dnf install --assumeyes python3-pip\n    pip3 install --upgrade pip\n  fi\n  pip3 install cargo-zigbuild\nfi",
            "cache_provider": "github"
          },
          {
            "runner": "ubuntu-22.04",
            "host": "x86_64-unknown-linux-gnu",
            "container": {
              "image": "quay.io/pypa/manylinux_2_28_x86_64",
              "host": "x86_64-unknown-linux-musl",
              "package_manager": null
            },
            "install_dist": {
              "shell": "sh",
              "run": "curl --proto '=https' --tlsv1.2 -LsSf https://github.com/axodotdev/cargo-dist/releases/download/v1.0.0-FAKEVERSION/cargo-dist-installer.sh | sh"
            },
            "dist_args": "--artifacts=local --target=aarch64-unknown-linux-musl",
            "targets": [
              "aarch64-unknown-linux-musl"
            ],
            "packages_install": "if ! command -v cargo-zigbuild > /dev/null 2>&1; then\n  if ! command -v pip3 > /dev/null 2>&1; then\n    dnf install --assumeyes python3-pip\n    pip3 install --upgrade pip\n  fi\n  pip3 install cargo-zigbuild\nfi",
            "cache_provider": "github"
          },
          {
            "runner": "macos-15-intel",
            "host": "x86_64-apple-darwin",
            "install_dist": {
              "shell": "sh",
              "run": "curl --proto '=https' --tlsv1.2 -LsSf https://github.com/axodotdev/cargo-dist/releases/download/v1.0.0-FAKEVERSION/cargo-dist-installer.sh | sh"
            },
            "dist_args": "--artifacts=local --target=x86_64-apple-darwin",
            "targets": [
              "x86_64-apple-darwin"
            ],
            "cache_provider": "github"
          },
          {
            "runner": "windows-2022",
            "host": "x86_64-pc-windows-msvc",
            "install_dist": {
              "shell": "pwsh",
              "run": "irm https://github.com/axodotdev/cargo-dist/releases/download/v1.0.0-FAKEVERSION/cargo-dist-installer.ps1 | iex"
            },
            "dist_args": "--artifacts=local --target=x86_64-pc-windows-msvc",
            "targets": [
              "x86_64-pc-windows-msvc"
            ],
            "cache_provider": "github"
          },
          {
            "runner": "ubuntu-22.04",
            "host": "x86_64-unknown-linux-gnu",
            "install_dist": {
              "shell": "sh",
              "run": "curl --proto '=https' --tlsv1.2 -LsSf https://github.com/axodotdev/cargo-dist/releases/download/v1.0.0-FAKEVERSION/cargo-dist-installer.sh | sh"
            },
            "dist_args": "--artifacts=local --target=x86_64-unknown-linux-gnu",
            "targets": [
              "x86_64-unknown-linux-gnu"
            ],
            "cache_provider": "github"
          },
          {
            "runner": "ubuntu-22.04",
            "host": "x86_64-unknown-linux-gnu",
            "install_dist": {
              "shell": "sh",
              "run": "curl --proto '=https' --tlsv1.2 -LsSf https://github.com/axodotdev/cargo-dist/releases/download/v1.0.0-FAKEVERSION/cargo-dist-installer.sh | sh"
            },
            "dist_args": "--artifacts=local --target=x86_64-unknown-linux-musl",
            "targets": [
              "x86_64-unknown-linux-musl"
            ],
            "packages_install": "sudo apt-get update\nsudo apt-get install musl-tools",
            "cache_provider": "github"
          }
        ]
      },
      "pr_run_mode": "plan"
    }
  },
  "linkage": [],
  "upload_files": [],
  "github_attestations": true
}

stderr:
